File: csc3100/A3/tree_sum.cpp
#include <iostream>
#include <vector>
#include <numeric>  

using namespace std;

const int MOD = 1000000007;

struct Node {
    vector<pair<int, int> > adj;
};

class BlackWhiteTree {
private:
    vector<Node> tree;
    vector<int> color;
    vector<long long> count, dis;
    vector<long long> ans;  
    long long total = 0;

public:
    BlackWhiteTree(int node_num) : tree(node_num), color(node_num), count(node_num),
                                   dis(node_num), ans(node_num) {}

    void saveColor(const vector<int>& colors) {
        color = colors;
        total = accumulate(color.begin(), color.end(), 0LL);
    }

    void addEdge(int u, int v, int w) {
        tree[u].adj.push_back(make_pair(v, w));
        tree[v].adj.push_back(make_pair(u, w));
    }

    void dfsHelper(int node, int parent) {
        count[node] = color[node];
        for (auto& adj_weight : tree[node].adj) {
            int adj = adj_weight.first;
            int weight = adj_weight.second;
            if (adj != parent) {
                dfsHelper(adj, node);
                count[node] += count[adj];
                dis[node] += dis[adj] + count[adj] * weight;
            }
        }
    }

    void dfs(int node, int parent) {
        ans[node] = dis[node];
        for (auto& adj_weight : tree[node].adj) {
            int adj = adj_weight.first;
            int weight = adj_weight.second;
            if (adj != parent) {
                long long node_to_adj_root = dis[node] - (dis[adj] + count[adj] * weight);
                long long node_to_adj = (total - count[adj]) * weight;
                dis[adj] += node_to_adj_root + node_to_adj;
                dfs(adj, node);
            }
        }
    }
    
    const vector<long long>& answer() {
        dfsHelper(0, -1);
        dfs(0, -1);
        return ans;
    }
};

int main() {
    int n;
    cin >> n;
    BlackWhiteTree tree(n);

    vector<int> colors(n);
    for (int i = 0; i < n; ++i) {
        cin >> colors[i];
    }
    tree.saveColor(colors);

    for (int i = 0; i < n - 1; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        tree.addEdge(u - 1, v - 1, w);
    }

    const vector<long long>& answers = tree.answer();
    for (int i = 0; i < n; ++i) {
        cout << (answers[i] % MOD) << endl;
    }

    return 0;
}

File: csc3100/A3/tree_sum.py
import sys
sys.setrecursionlimit(2000000)
MOD = 10**9 + 7

class Node():
    def __init__(self):
        self.adj = []

class blackWhiteTree():
    def __init__(self, node_num):
        self.tree = [Node() for i in range(node_num)]
        self.color = [0] * node_num
        self.count = [0] * node_num
        self.dis = [0] * node_num
        self.ans = [0] * node_num
        self.total = 0  
    
    def save_color(self, color):
        self.color = color
        self.total = sum(self.color) 

    def save_node(self, u, v, w):
        self.tree[u].adj.append((v, w))
        self.tree[v].adj.append((u, w))
    
    def dfs_helper(self, node, pt):
        self.count[node] = self.color[node]
        for adj, weight in self.tree[node].adj:
            if adj != pt:
                self.dfs_helper(adj, node)
                self.count[node] += self.count[adj]
                self.dis[node] += self.dis[adj] + self.count[adj] * weight
    
    def dfs(self, node, pt):
        self.ans[node] = self.dis[node]
        for adj, weight in self.tree[node].adj:
            if adj != pt:
                prev_dis_adj = self.dis[adj]
                self.dis[adj] += self.dis[node] - (prev_dis_adj + self.count[adj] * weight) + \
                    (self.total - self.count[adj]) * weight
                self.dfs(adj, node)

    
    
n = int(input())
tree = blackWhiteTree(n)

color = list(map(int, input().split()))
tree.save_color(color)

for _ in range(n - 1):
    u, v, w = map(int, input().split())
    tree.save_node(u - 1, v - 1, w)

tree.dfs_helper(0, -1)
tree.dfs(0, -1)

for i in tree.ans:
    print(i % MOD)
